### PR TITLE
Adding advanced networking, additional hardware, and more

### PR DESCRIPTION
**Pull request by marc-ta**
Added several features here:
- Attribute to specify vApp name
- Advanced networking
  - Add Org VDC networks to vApp
  - Add vApp networks to vApp
- Add vApp metadata
- Add VM metadata
- Guest customization rolled up into compose/recompose to avoid additional call
- Moved set_hardware to build_vapp from power_on, it seemed to make more sense there
- Add hard disks
- Update / add network cards and configuration
- Added switches for power_on, ssh, file syncing, and auto uploading boxes to vCloud

### CODE REVIEW

**Comment by StefanScherer** [comment]
@marc-ta thanks again for the add_hdds feature. I've just cleaned up my [basebox-slave](https://github.com/StefanScherer/basebox-slave) installation with [this commit](https://github.com/StefanScherer/basebox-slave/commit/862ee1e8f2638d69752dac147b643ee154f5ff07) to use default sized base boxes (Windows) for the Jenkins slaves and adding a second disk and mount it as my Jenkins work directory. 

I'm currently running a pre-release of vagrant-vcloud with your PR and some of mine.
BTW: The vagrant-serverspec plugin is going to Serverspec V2, I've started to use it to test my base boxes.

**Comment by StefanScherer** [comment]
@marc-ta Seems that needs a rebase.

**Comment by StefanScherer** [comment]
@frapposelli What about merging this after a rebase? This is a major enhancement for the vagrant-vcloud plugin. A colleage just asked me for the change SID feature.

File: lib/vagrant-vcloud/action/forward_ports.rb [hunk 1]
```diff
@@ -64,24 +70,26 @@ def forward_ports
...
-                :nat_policy_type  => 'allowTraffic',
-                :nat_rules        => ports
-              }
-            )
-
-            wait = cnx.wait_task_completion(add_ports)
-
-            if !wait[:errormsg].nil?
-              raise Errors::ComposeVAppError, :message => wait[:errormsg]
+            ports.values.each do |port|
```
Comment (stefanschererseal) [inline lib/vagrant-vcloud/action/forward_ports.rb]: You have changed the `ports` from an array to a hash. Here it works fine.

But in [line 108](https://github.com/frapposelli/vagrant-vcloud/pull/104/files#diff-249abf56cedf2792aee220386a4932bfR108) when using an edge gateway the code still accesses the "array" `ports` and crashes there.

Comment (stefanschererseal) [inline lib/vagrant-vcloud/action/forward_ports.rb]: @marc-ta Changing line 108 from `ports.each` to `ports.values.each` seems to fix the problem with the edge gateway. 

Could you please update this and rebase? :heart:

**Comment by frapposelli** [comment]
Sure, it's definitely ripe for merge, thanks for doing CR on this @StefanScherer 

Once the rebase and the fixes are done I'll merge on dev.

**Comment by StefanScherer** [comment]
One change is missing, but I have sent a PR to fix this PR in the forked repo.

**Comment by marc-ta** [comment]
@StefanScherer, @frapposelli 
I have rebased, added the pull request, and added a few fixes that I found I hadn't pushed up.

**Comment by frapposelli** [comment]
@marc-ta looks good, can you please squash the commits and force push?

@StefanScherer did you test it? everything looks good on your side?

**Comment by StefanScherer** [comment]
@frapposelli I tried a small test from a clean Windows box. But I've installed Vagrant 1.8.1 and now I get an error. I've opened #122 to have a closer look at that. 
Now setting up another test env with Vagrant 1.7.4 and have a test drive...

**Comment by StefanScherer** [comment]
@frapposelli Down to Vagrant 1.7.4 I have tested the gem with the latest commits of this PR. It creates the edge gateway rules. LGTM.

**Comment by frapposelli** [comment]
LGTM

**Comment by frapposelli** [comment]
whops, just realized this was sent against `master`, can you please resend this against `develop` ?

**Comment by marc-ta** [comment]
moving to dev branch

**Comment by marc-ta** [comment]
New PR #123 added.

**Pull request closed without merging**